src/v/cloud_roles/refresh_credentials.cc

```diff
@@ -0,0 +1,289 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_roles/refresh_credentials.h"
+
+#include "cloud_roles/aws_refresh_impl.h"
+#include "cloud_roles/aws_sts_refresh_impl.h"
+#include "cloud_roles/gcp_refresh_impl.h"
+#include "cloud_roles/logger.h"
+#include "config/node_config.h"
+#include "model/metadata.h"
+#include "s3/client.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/loop.hh>
+
+/// These environment variables can be used to override the default hostname and
+/// port for fetching temporary credentials for testing.
+struct override_api_endpoint_env_vars {
+    static constexpr const char* host = "RP_SI_CREDS_API_HOST";
+    static constexpr const char* port = "RP_SI_CREDS_API_PORT";
+};
+
+/// Multiplier to derive sleep duration from expiry time. Leaves 0.1 * expiry
+/// seconds as buffer to make API calls. For default expiry of 1 hour, this
+/// results in a fetch after 54 minutes.
+static constexpr float sleep_from_expiry_multiplier = 0.9;
+
+cloud_roles::refresh_credentials::refresh_credentials(
+  std::unique_ptr<impl> impl,
+  ss::gate& gate,
+  ss::abort_source& as,
+  credentials_update_cb_t creds_update,
+  s3::aws_region_name region)
+  : _impl(std::move(impl))
+  , _gate(gate)
+  , _as(as)
+  , _credentials_update(std::move(creds_update))
+  , _region{std::move(region)} {}
+
+void cloud_roles::refresh_credentials::start() {
+    ssx::background = ssx::spawn_with_gate_then(_gate, [this]() mutable {
+        return ss::do_until(
+          [this] { return _gate.is_closed() || _as.abort_requested(); },
```

Comment:
@Lazin should I check only that abort source is triggered here? This object shares the abort source and gate with remote,  and remote will trigger abort source on stop. 

I added the gate check here but now I think it will never be required and just abort source check is enough.